### PR TITLE
Block Editor: Consider events only from DOM descendents (WritingFlow, ObserveTyping)

### PR DESCRIPTION
**Pull request by aduth**
Alternative to: #19804

This pull request seeks to improve upon how events are interpreted by WritingFlow and ObserveTyping. Historically, we have often resorted to `Event#stopPropagation` within many components for the sole purpose of avoiding this top-level event handling, but this is problematic for a number of reasons:

- It is not obvious to implement this, and is usually added later only once the undesirable behavior has been identified.
   - Example: #11088
- `Event#stopPropagation` is an extreme solution and often introduces a lot of unintended incompatibilities and interoperability issues. It should be avoided whenever possible.
  - Related reading: https://css-tricks.com/dangers-stopping-event-propagation/
- It introduces an undesirable implicit dependency between the component and its ancestors
  - Example: In stopping propagation, `URLInput` (`@wordpress/components`) establishes a dependency on both its editor ancestry (`@wordpress/block-editor` `WritingFlow`, `ObserveTyping`) and an assumption that it is shown in [`Popover`](https://github.com/WordPress/gutenberg/tree/master/packages/components/src/popover), despite the fact it is intended to be a general-purpose component.

In some of the discussion from https://github.com/WordPress/gutenberg/pull/19804#discussion_r369891119 , it is considered that it might be possible to consider the various dialog components (`Popover`, `Modal`) as entirely separate contexts and to not allow events to bubble through to its ancestor components. In some ways, this aligns with its implementation using Slot/Fill, where the intention of rendering the dialogs elsewhere in the DOM is largely to treat it as being positioned independently from its ancestors (in practical terms, to avoid CSS cascading styles).

The implementation here achieves this by considering whether the events handled by ObserveTyping and WritingFlow originate from within its own DOM hierarchy. This is meant to be distinguished from event bubbling which occurs through the React hierarchy, and provides an opportunity for any components which do not want to be subject to this handling to avoid detection by rendering elsewhere in the DOM. In the case of popovers, this happens by virtue of the fact that they are rendered using [`Slot` and `Fill` components](https://github.com/WordPress/gutenberg/tree/master/packages/components/src/slot-fill).

**Testing Instructions:**

Verify that affected components continue to work as expected, notably that:

- When typing in a paragraph, the block contextual toolbar is dismissed until the mouse is moved, Escape is pressed, or a non-collapsed selection is made (ObserveTyping behaviors)
- Arrow keys traverse across blocks (WritingFlow behaviors)
- These same interactions are not affected when typing occurs within a popover
   - Example: When adding a link to a paragraph, the block contextual toolbar should remain visible while typing within the link editor.

### CODE REVIEW

**Comment by aduth** [comment]
To simplify review here, I extracted the function-component refactoring of `ObserveTyping` (3e0159abf26bc94991a5d626561ec29169522ef2) to its own pull request at #19881.

**Comment by ellatrix** [comment]
This is great. I’ll review tomorrow if I find some time

File: packages/block-editor/src/components/link-control/search-input.js [hunk 1]
```diff
@@ -4,38 +4,12 @@
...
-	}
-};
-
-const handleLinkControlOnKeyPress = ( event ) => {
-	const { keyCode } = event;
-
-	event.stopPropagation();
-
-	if ( keyCode === ENTER ) {
-
```
Comment (ellatrix) [inline packages/block-editor/src/components/link-control/search-input.js]: Huh

**Comment by aduth** [comment]
The failing end-to-end tests could be legitimate, based on the error message, since the new code checks DOM descendent by `Element#contains`:

```
● autosave › should save to sessionStorage
    TypeError: Cannot read property 'contains' of undefined
```

Not sure how the `ref.current` could be `undefined` at the point that it's handling `onKeyDown`, `onFocus`, `onKeyPress`. At worst, we could just add an extra check there to ensure it's truthy, but it's worth investigating why it's happening in the first place.

File: packages/block-editor/src/components/link-control/search-input.js (hunk 1 above)
Comment (aduth) [reply packages/block-editor/src/components/link-control/search-input.js]: 🤷‍♂ https://github.com/WordPress/gutenberg/pull/19458#discussion_r367600295

**Comment by aduth** [comment]
I've rebased this after merging #19881, so the `ObserveTyping`-specific changes should be fewer.

The test failure was quite bizarre. From debugging, the [`node.current.contains()` condition](https://github.com/WordPress/gutenberg/blob/71af6b94a33c1bf76c0638439822fd074403715e/packages/block-editor/src/components/observe-typing/index.js#L75) was being triggered in response to a focus event on the initial render, caused by the auto-focus of the post title. But it's strange to me that this event handler could ever be called without the ref also being set, considering that [both should take effect in the same render](https://github.com/WordPress/gutenberg/blob/71af6b94a33c1bf76c0638439822fd074403715e/packages/block-editor/src/components/observe-typing/index.js#L189-L190). It could be a timing issue where event handlers are added before the ref is assigned, and if those event handlers happen to be triggered in the meantime, there's a rare chance the ref may still be unassigned.

In any case, it occurred to me there's a better implementation here anyways, avoiding the ref altogether and simply using [event.currentTarget](https://developer.mozilla.org/en-US/docs/Web/API/Event/currentTarget), which should be effectively the same as what we're wanting, but more closely associated with the event and the DOM, and is slightly simpler code.

**Comment by aduth** [comment]
This isn't viable without changing how we approach navigation mode tabbing:

https://github.com/WordPress/gutenberg/blob/9e3c713f5afab9a558b86b3daefd233621dbad36/packages/block-editor/src/components/writing-flow/index.js#L325-L354

The problem is: The <kbd>Tab</kbd> event originates from the toolbar button:

image

...which is now rendered in a Popover as of #19564. And these are the exact sorts of events we're otherwise wanting to ignore.

The issue can be reproduced following this test case:

https://github.com/WordPress/gutenberg/blob/9e3c713f5afab9a558b86b3daefd233621dbad36/packages/e2e-tests/specs/editor/various/keyboard-navigable-blocks.test.js#L107-L134

Essentially, the <kbd>Tab</kbd> might skip over the logic to focus the next block's label, and instead circle back to the title.

A few thoughts on how to approach this:

- Move out navigation mode handling to a separate component which _will_ handle React bubbled events.
- Render the [toolbar-specific Slot](https://github.com/WordPress/gutenberg/blob/9e3c713f5afab9a558b86b3daefd233621dbad36/packages/edit-post/src/components/visual-editor/index.js#L32) within `WritingFlow` so it can be considered as part of the same DOM.
   - On first glance, it seems this could work, but in truth I'm not really sure what the original purpose of this custom slot is meant to be, and if there are other similar issues we might encounter with other non-toolbar Popover renderings.
- Find a way to allow native tab behavior to handle block navigation.
   - e.g. Assuming that the next block could receive focus for the tab event, maybe it can be the responsibility of the block to show/focus the navigation label when it receives focus.
- Revert #19564.
   - I assume want to avoid this, but for exhaustiveness' sake, it is technically an option.

**Comment by ellatrix** [comment]
@aduth Trying to understand the issue here: is the problem that the breadcrumb button is not part of the DOM tree or that it is not part of the React tree?

**Comment by aduth** [comment]
> @aduth Trying to understand the issue here: is the problem that the breadcrumb button is not part of the DOM tree or that it is not part of the React tree?

It's a problem because it's not part of the DOM tree (it is not a DOM descendent of WritingFlow wrapper).

**Comment by ellatrix** [comment]
I see two options.

* Either we move the navigation mode logic to the breadcrumb button component (focus stays on the button during navigation, only selection changes) and we skip the content when you reach the end (the breadcrumb is located above the content).
* Or we do not set focus on button but rather on the blocks.

I find the latter more intuitive since when you're navigating content, focus also moves from block to block. This also ensures that all logic is kept within the `WritingFlow` component. With option one I fear that all the logic will become fragmented.

I don't really understand why focus is set on a button in navigation mode right now. @youknowriad @MarcoZehe do you remember the reasons behind that?

**Comment by youknowriad** [comment]
> Or we do not set focus on button but rather on the blocks.

This is not an option for navigation mode because it breaks to have keyboard event handlers on non-button elements (div) in some windows screen readers (NVDA I think). That's the main reason the navigation mode was added a bit late.

**Comment by ellatrix** [comment]
> This is not an option for navigation mode because it breaks to have keyboard event handlers on non-button elements (div) in some windows screen readers (NVDA I think).

That's strange. We currently already have  keyboard event handlers on `div` elements (enter and delete/backspace on the block wrapper in edit mode). Does that not work then?

**Comment by youknowriad** [comment]
>  Does that not work then?

Probably not unless that element has a specific role. The specificity here is that this feature was specifically dedicated to a11y users so it didn't make sense for these handlers to not work on all screen readers.

File: packages/block-editor/src/components/link-control/search-input.js (hunk 1 above)
Comment (getdave) [reply packages/block-editor/src/components/link-control/search-input.js]: I noticed this and removed it in https://github.com/WordPress/gutenberg/pull/19775.

Not sure how it happened. Sorry!

**Comment by ellatrix** [comment]
In #27489, I wrote the missing piece for this PR: moving navigation logic from writing flow to the block selection button (the focussed element in navigation mode). Previously I also rewrote the typing observer as a hook in #27043 (which now only listens for events of DOM descendants). @aduth, thanks for creating this PR. :)

**Pull request closed without merging**